File: 241-different-ways-to-add-parentheses/ken/241.py
class Solution(object):
    def diffWaysToCompute(self, input):
        """
        :type input: str
        :rtype: List[int]
        """

        if '+' not in input and \
           '-' not in input and \
           '*' not in input:
           return [int(input)]

        ans = []

        for i in xrange(len(input)):
            c = input[i]

            if c != '+' and c != '-' and c != '*':
                continue

            v1 = self.diffWaysToCompute(input[:i])
            v2 = self.diffWaysToCompute(input[i+1:])

            for a in v1:
                for b in v2:
                    if c == '+':
                        ans.append(a + b)
                    elif c == '-':
                        ans.append(a - b)
                    elif c == '*':
                        ans.append(a * b)

        return ans

File: 241-different-ways-to-add-parentheses/leochang/241.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string input) {
        vector<int> ans;
        for (int i = 0; i < input.size(); i++) {
            if (input[i] == '+' || input[i] == '-' || input[i] == '*') {
                vector<int> ans1 = diffWaysToCompute(input.substr(0, i));
                vector<int> ans2 = diffWaysToCompute(input.substr(i+1));
                for (auto c1 : ans1) {
                    for (auto c2 : ans2) {
                        if (input[i]=='+') ans.push_back(c1+c2);
                        else if (input[i]=='-') ans.push_back(c1-c2);
                        else if (input[i]=='*') ans.push_back(c1*c2);
                    }
                }
            }
        }

        return ans.size() ? ans : vector<int>{stoi(input)};
    }
};
